tests/trace/test_iterators.py

```diff
@@ -0,0 +1,142 @@
+import threading
+
+import pytest
+
+from weave.utils.iterators import ThreadSafeInMemorySequence
+
+
+def test_basic_sequence_operations():
+    # Test basic sequence operations
+    iterator = ThreadSafeInMemorySequence(iter(range(10)))
+    assert len(iterator) == 10
+    assert iterator[0] == 0
+    assert iterator[1:3] == [1, 2]
+    assert list(iterator) == list(range(10))
+
+
+def test_empty_iterator():
+    # Test behavior with empty iterator
+    iterator = ThreadSafeInMemorySequence(iter([]))
+    assert len(iterator) == 0
+    with pytest.raises(IndexError):
+        _ = iterator[0]
+    assert list(iterator) == []
+
+
+def test_known_length():
+    # Test initialization with known length
+    iterator = ThreadSafeInMemorySequence(iter(range(5)), known_length=5)
+    assert len(iterator) == 5  # Should not need to exhaust iterator
+    assert iterator[4] == 4  # Access last element
```

Comment:
not sure this test does what you think it does

I think even if you exhausted the iterator, this would still be possible because you're caching the values, right?

Reply:
true, but its more to clarify the intention of the sequence, where asking for length is a non-destructive/iterative operation